### PR TITLE
mimic apache resolves #12

### LINKED ISSUE #12: Mimic Apache

### PR DESCRIPTION
**Pull request by fffw**
Resolves #12 

It compares the response of real Apache web server and chained server byte to byte for a set of URL patterns and methods to ensure we mimic it accurately.

### CODE REVIEW

File: main_test.go [hunk 1]
```diff
@@ -593,26 +593,26 @@ type proxy struct {
 func setupNewHTTPServer(maxConns uint64, idleTimeout time.Duration) (*Server, error) {
 	s := NewServer(validToken, maxConns, idleTimeout, false, utils.QUIET)
 	var err error
-	ready := make(chan bool)
+	chListenOn := make(chan string)
```
Comment (fffw) [inline main_test.go]: Naming is always a headache to me, for example, this ugly `chListenOn`

File: mimic/apache.go [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+type apacheMimic struct {
+	conn net.Conn
+	req  *http.Request
+	path string
+}
+
+var (
+	Host         string
+	Port         string
+	lastModified = time.Now().Format("Fri, 22 Oct 2015 11:52:25 GMT")
```
Comment (fffw) [inline mimic/apache.go]: Weird things happen on my go1.5.1, the formatted string becomes something like `Fri, 3030 Oct 301052 1010:5230:3052 GMT`, hope it's a new introduced bug and not in previous versions.

Comment (fffw) [inline mimic/apache.go]: Ok, find a workaround and submit an issue at https://github.com/golang/go/issues/13103

**Comment by fffw** [comment]
There are a few failed tests (main_test.go, L112, L143, L235 and L247), but the related test code is a bit obscure to me. Still trying to fix them.

BTW, the screen output is overwhelming with `-debug` option. It would be good to have another level of logging for http dumping. Can we add something like `DumpRequest()` and `DumpResponse()` to `golog` and switch to use it?

File: main_test.go (hunk 1 above)
Comment (uaalto) [inline main_test.go]: Naming is one of the most important things in my opinion. I also stop and think about names a lot...

**Comment by uaalto** [comment]
So those test are failing now?
We need to be able to set logging *per middleware*, besides the particular case of dumping requests only if the level is above the one set. Creating DumpRequest/Response is already two methods instead of one (IsLevel), and they solve only that particular situation.
I don't think Golog provides any key functionality, and it is always going to be biased towards the Lantern client. Why are you interested in using it?

**Comment by fffw** [comment]
Yeah I added travis integration so you can also see the failed test there.

What I want is to change logging behaviour without change any line of code. For example, I don't want to dump requests/responses with `-debug` option, but want to see other debug outputs. Or show detailed output for any single package with an option or environment variable.

In `golog`, I can `TRACE=detour,balancer ./lantern` to see trace logs of the two packages but not others, which also makes sense for server debugging.

Anyway, it's just for convenience, so either one is ok.

**Comment by uaalto** [comment]
Ok, that's a change that definitely makes sense. That kind of control is very useful.
What do you think about implementing that here? The reason I would like to keep them separate is because I think we are going to diverge in the needs of logging for the server and the client (inherently different problems), so we can do changes here without affecting the client. What do you think?

File: devicefilter/devicefilter.go [hunk 3]
```diff
@@ -52,8 +53,8 @@ func (f *DeviceFilter) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	lanternDeviceId := req.Header.Get(deviceIdHeader)
 
 	if lanternDeviceId == "" {
-		f.log.Debugf("No %s header found, respond 404 not found to %s\n", deviceIdHeader, req.RemoteAddr)
-		w.WriteHeader(http.StatusNotFound)
+		f.log.Debugf("No %s header found from %s, mimicking apache\n", deviceIdHeader, req.RemoteAddr)
+		mimic.MimicApache(w, req)
```
Comment (uaalto) [inline devicefilter/devicefilter.go]: A suggestion on naming, really minor: I would say "MimicApacheError".

File: mimic/apache.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+		vars.ContentLength = bodyBuf.Len()
+	}
+	err := header.Execute(&headBuf, vars)
+	if err != nil {
+		panic(fmt.Sprintf("execute template err: %s", err))
+	}
+	// ignore any errors writing back to connection
+	_, _ = headBuf.WriteTo(f.conn)
+	_, _ = bodyBuf.WriteTo(f.conn)
+	_ = f.conn.Close()
```
Comment (uaalto) [inline mimic/apache.go]: I think you put this here because deferring is more costly, is that right?
I don't think is a problem, because we are panicking on errors, but since those panics happen in goroutines the program might not terminate and leave this connections open. That said, if the goroutine panics the templates are buggy and need a server fix anyway.

File: devicefilter/devicefilter.go (hunk 3 above)
Comment (uaalto) [inline devicefilter/devicefilter.go]: I read below in the comments that it actually mimics an unconfigured Apache. Should we convey that in the naming?

File: mimic/apache.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+		Host:         Host,
+		Port:         Port,
+	}
+}
+
+func makeETag() string {
+	const alphanum = "0123456789abcdefghijklmnopqrstuvwxyz"
+	bytes := [17]byte{}
+	rand.Read(bytes[:])
+	for i, b := range bytes {
```
Comment (uaalto) [inline mimic/apache.go]: Sorry for the nitpicking... why don't you run the second case and outside the range looping just set:
```
bytes[4] = '-'
```

File: mimic/index-dot-html.go [hunk 6]
```diff
@@ -0,0 +1,966 @@
+// File generated by 2goarray v0.1.0 (http://github.com/cratonica/2goarray)
+
+package mimic
+
+var indexDotHTML []byte = []byte{
```
Comment (uaalto) [inline mimic/index-dot-html.go]: What is the main reason for embedding this instead of loading it from file and keeping in memory? I'm asking because you took the time to do this, so you probably saw a compelling reason.

File: mimic/apache.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+	if err != nil {
+		panic("fail to hijack, should not happen")
+	}
+	path := req.URL.Path
+	// remove extra leading slash
+	if len(path) > 0 && path[0] == '/' {
+		i := 1
+		for ; i < len(path) && path[i] == '/'; i++ {
+		}
+		if i > 0 {
```
Comment (uaalto) [inline mimic/apache.go]: Isn't `i` going to be always greater than 0?

File: mimic/apache.go [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+	}
+	m := apacheMimic{conn, req, path}
+	if req.Host == "" {
+		m.writeError(badRequestHeader, badRequestBody)
+		return
+	}
+	switch req.Method {
+	case "CONNECT":
+		m.writeError(badRequestHeader, badRequestBody)
+	case "GET", "POST":
```
Comment (uaalto) [inline mimic/apache.go]: Is Apache responding to a POST to an image path?

File: main_test.go (hunk 1 above)
Comment (uaalto) [inline main_test.go]: The name is fine in my opinion.

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+language: go
+go:
+- 1.4.3
```
Comment (uaalto) [inline .travis.yml]: Oh, I was using Go 1.5.1 to develop this, since it has the better-performing GC which probably benefits the proxy. It apparently does run in both versions though.

**Comment by uaalto** [comment]
Thanks @fffw. I finished my first round of the review.

**Comment by fffw** [comment]
Given we already invested a lot of time on golog, I can't think of an
reason to maintain another logging utility. Golog has no client specific
feature AFAIK, and fronted server also use it. I think we can add IsLevel
call to golog and everything are done.

File: devicefilter/devicefilter.go (hunk 3 above)
Comment (fffw) [inline devicefilter/devicefilter.go]: It feels to me not an error but the behavior of Apache, as it can also
respond 200 OK

File: mimic/apache.go (hunk 4 above)
Comment (fffw) [inline mimic/apache.go]: Good catch! I just wrote it here and didn't notice the potential leak.

File: devicefilter/devicefilter.go (hunk 3 above)
Comment (fffw) [inline devicefilter/devicefilter.go]: If we can find a name with comparable length ;)

Comment (uaalto) [inline devicefilter/devicefilter.go]: Ok.

**Comment by uaalto** [comment]
Well, Golog is about 200 lines of code, not much to replicate...
Worst case scenario if we bring that dependency is that it needs to make a different version at some point. However, I still see the optSetter pattern as an important one, especially since we are going to provide a "standard" HTTP proxy with compatibility with other providers of middleware (such as Oxy).

**Comment by uaalto** [comment]
I've taken a look at the failing tests. Which are the ones you find obscure? I'm guessing that max connections and idle connections, the rest are just request/response round trips.

File: mimic/apache.go (hunk 7 above)
Comment (fffw) [inline mimic/apache.go]: Thanks! I added the outer check but forgot to remove the inner on.

File: .travis.yml (hunk 9 above)
Comment (fffw) [inline .travis.yml]: I copied the file from lantern repo and tweaked it a bit, good to change it to newest version.

File: mimic/apache.go (hunk 8 above)
Comment (fffw) [inline mimic/apache.go]: Thanks for reminding me. Added checks in test.

File: mimic/index-dot-html.go (hunk 6 above)
Comment (fffw) [inline mimic/index-dot-html.go]: Extra file brings lots of extra work. Checking its existence and integrity , copying them everywhere, etc. Embedding everything is also what Go does and it‘s a very good feature.

**Comment by fffw** [comment]
`golog` doesn't conflict with `optSetter` pattern as long as it implements `utils.Logger` interface (or, make an adapter if there's a solid reason to not change `utils.Logger` interface).

**Comment by fffw** [comment]
I tweaked the test to reflect the actual errors we want to capture, but still don't know where those EOFs and timeouts come from...

**Comment by fffw** [comment]
@uaalto updated code as per your suggestions, though still need some time to fix those failed tests.

**Comment by uaalto** [comment]
OK, great!
The Max connections and Idle timeout tests rely on timings, if they become imprecise then it might be better to use `fdcount` for that, although in principle they should be working.

**Comment by uaalto** [comment]
Also, if you are interested, we can also move to golog. I don't find the Logger option that important if we have a way to specify the modules that we want to log verbose via environment variables. I actually agree that via environment variables is more flexible than by code for this purpose.

File: mimic/apache.go (hunk 5 above)
Comment (fffw) [inline mimic/apache.go]: either way is ok to me

**Comment by fffw** [comment]
The `EOF` error previous test captures is what `Read` returns after client closes connection. `mimic` package changes this behaviour by forcibly closing connection at server side. It's a violation of HTTP 1.1 standard as server should only close connection if "Connection:close" header is sent back to client （section 14.10 of rfc2616).

`mimic` has to do so because we can't further process an hijacked connection, and the reason why we hijack it is to make sure the response headers are of same order as Apache.

@uaalto Do you think it's ok to alter those failed asserts in `TestReportStats` and `TestIdleTargetConnections` to reflect this change of behaviour?

The failed asserts in `TestMaxConnections` is also because `mimic` closed connection immediately. I changed it to send request with valid tokens to avoid hitting `mimic` (which is also more meaningful as the limitation is mainly applied to valid requests).

**Comment by uaalto** [comment]
I see. Well we have no other option, but to reimplement the whole Go server we are based on. I suspected that at some point the order of header fields was going to be an issue, given the degree of control we need here. They do matter in some cases.
No problem in changing those tests. The idle connection tests are "integration" tests, in the sense that measure the whole request lifetime. Idletiming is already being thoroughly unit-tested as an independent package, so I'm not very worried about that test.

I'm thinking, maybe we should evaluate the possibility of copying and customizing the Go server. I have found a few corner cases form Co Advisor that expose tiny bug in the depths of the server, and I'm not sure if is possible to do it without touching there.

Anything that requires hijacking is already requiring a fully-manual approach, which I think fares worse than making our own custom server.

**Comment by fffw** [comment]
Finally passed (though more appropriate word would be 'cheated') all tests. @uaalto Could you take a look again?

Building `http-proxy` directly on top of TCP and TLS gives us fine control as well as potential performance boost, but consider the substantial workload to implement and, more importantly, maintain the code, I would postpone discussing that option until Lantern pro has already stabilized.

**Comment by uaalto** [comment]
Sure, I'm well aware of that, but some of the issues are very hard to implement otherwise, that's why I mention it. If we don't care about those, then we are fine.

**Comment by fffw** [comment]
It is also recognized by `nmap` as Apache server.
> 8080/tcp open     http         Apache httpd 2.4.7 ((Ubuntu))

**Comment by uaalto** [comment]
That's nice, I was going to suggest that too. One thing, as an even more precise mimicking that we could do, is to run an unconfigured Apache through Co Advisor and then try to duplicate exactly the same success/failure/violation results. I don't think we need that yet though.

**Comment by uaalto** [comment]
It's actually a lot of work...

**Comment by uaalto** [comment]
By the way, we can talk about that later in the meetup, but actually the order of header fields is important for the meaning of some fields, and in some tests (such as those in chunking) they don't pass for that reason after my fixes.

**Comment by uaalto** [comment]
LGTM!
